crates/hir-def/src/signatures.rs

```diff
@@ -889,21 +891,53 @@ fn lower_fields<'a>(
     (arena, store, col.source_map)
 }
 
+#[derive(Debug, PartialEq, Eq)]
+pub struct InactiveEnumVariantCode {
+    pub cfg: CfgExpr,
+    pub opts: CfgOptions,
+    pub ast_id: span::FileAstId<ast::Variant>,
+}
+
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct EnumVariants {
     pub variants: Box<[(EnumVariantId, Name)]>,
 }
+
 impl EnumVariants {
-    pub(crate) fn enum_variants_query(db: &dyn DefDatabase, e: EnumId) -> Arc<EnumVariants> {
+    pub(crate) fn enum_variants_query(
+        db: &dyn DefDatabase,
+        e: EnumId,
+    ) -> (Arc<EnumVariants>, Arc<ThinVec<InactiveEnumVariantCode>>) {
         let loc = e.lookup(db);
         let item_tree = loc.id.item_tree(db);
 
-        Arc::new(EnumVariants {
-            variants: loc.container.def_map(db).enum_definitions[&e]
-                .iter()
-                .map(|&id| (id, item_tree[id.lookup(db).id.value].name.clone()))
-                .collect(),
-        })
+        let mut diagnostics = ThinVec::new();
+        let cfg_options = loc.container.krate.cfg_options(db);
+        let mut index = 0;
+        let variants = FileItemTreeId::range_iter(item_tree[loc.id.value].variants.clone())
+            .filter_map(|variant| {
+                let attrs = item_tree.attrs(db, loc.container.krate, variant.into());
+                if attrs.is_cfg_enabled(cfg_options) {
+                    let enum_variant = EnumVariantLoc {
+                        id: ItemTreeId::new(loc.id.tree_id(), variant),
+                        parent: e,
+                        index,
+                    }
+                    .intern(db);
+                    index += 1;
+                    Some((enum_variant, item_tree[variant].name.clone()))
+                } else {
+                    diagnostics.push(InactiveEnumVariantCode {
+                        ast_id: item_tree[variant].ast_id,
+                        cfg: attrs.cfg().unwrap(),
+                        opts: cfg_options.clone(),
+                    });
+                    None
+                }
+            })
+            .collect();
+
+        (Arc::new(EnumVariants { variants }), Arc::new(diagnostics))
```

Comment:
Should probably `shrink_to_fit()` the diagnostics, also return `Option<Arc<ThinVec>>` to avoid allocation when there aren't any diagnostics (ideally this will return a reference then we could return just a `ThinVec`, but not for now).